parsely/src/main/java/com/parsely/parselyandroid/LocalStorageRepository.kt

```diff
@@ -33,23 +32,21 @@ internal open class LocalStorageRepository(private val context: Context) {
         }
     }
 
-    /**
-     * Delete the stored queue from persistent storage.
-     */
-    fun purgeStoredQueue() {
-        persistObject(ArrayList<Map<String, Any>>())
-    }
+    open suspend fun remove(toRemove: List<Map<String, Any?>?>) {
```

Comment:
_This comment is not about this specific line, but I think this is the best place available in this PR review._

There is a slight inconsistency about where the Mutex lock is applied within this class. It seems every function has the lock applied, except for `persistObject`. However, `persistObject` is also always called from within a lock.

Also, it seems the mutex lock is sometimes called from an already locked state as such:
```
mutex.withLock {
  mutex.withLock {
  }
}
```
I am not entirely sure how this is handled internally, but at the very least, it seems a bit inefficient. This inefficiency might be necessary in order to keep things safe, I am not sure.

I think it'd be good to look into how Mutex lock is used as a whole to make it more consistent and possibly more efficient. But, maybe it could be done in its own PR?

Reply:
Thanks for this comment. I'm totally fine with changing this logic, but I'd like to explain my rationale on those decisions.

> It seems every function has the lock applied, except for `persistObject`.

True, and it's intended. As you rightfully notice, `mutex` of `LocalStorageRepository` locks execution of all public methods. As we control all possible executions of `private persistObject`, it's safe to leave it without `mutex`. Also, adding the same lock to this method would create a thread lock ([sample](https://pl.kotl.in/p9DbWDKUb)).

> Also, it seems the mutex lock is sometimes called from an already locked state as such:

It's crucial to emphasize, that in the SDK those two `mutex`es are different objects and have different purpose. If it was the same object, we would have a thread lock (sample above). Adding more context to the example you provided:

```
mutexA.withLock { (1)
  mutexB.withLock { (2)
        persistObject
  }
  //continue with execution locked by `mutexA` only
}
```

`mutexA` is from `SendEvents`/`FlushQueue` use case. It doesn't allow running this use case twice at the same time. Otherwise, if we run it twice, we're at risk of sending duplicated events (first execution wouldn't call `remove` method before second execution called `getStoredQueue`). I could address it in implementation of the use case by e.g. introducing `LocalStorageRepository#getAndRemoveStoredQueue` but I thought `Mutex` is also a fine solution.

Also, you might want to take a look at `InMemoryBuffer` - in case of this class, the `mutexA` is there, but the reasoning is the same (although I don't have an idea for a different, no-mutex approach there)

`mutexB` is from `LocalStorageRepository`. It doesn't allow manipulating the local file at the same time by two different methods. It's crucial to not risk corrupting the local file and safely performing operations.

> I think it'd be good to look into how Mutex lock is used as a whole to make it more consistent and possibly more efficient. But, maybe it could be done in its own PR?

Sure - we could probably remove `mutexA` (I mean: the one in `SendEvents`/`FlushQueue` class) but I'm not sure if it's crucial for performance. If you see other improvements, I'm eager to discuss them! If you'd like to experiment with `Mutex` usage, I can recommend running `FunctionalTests#stressTest` for verification if coroutines are not locking and/or are efficient.

Reply:
> It's crucial to emphasize, that in the SDK those two `mutex`es are different objects and have different purpose. If it was the same object, we would have a thread lock (sample above). Adding more context to the example you provided:

Yeap, I was aware of that, and I should have specified that in my pseudocode as I can see how it could be interpreted as the same locks. Sorry about that!

> True, and it's intended. As you rightfully notice, `mutex` of `LocalStorageRepository` locks execution of all public methods. As we control all possible executions of `private persistObject`, it's safe to leave it without `mutex`.

I think in the context of "some of these functions are meant to be publicly consumed", trade-off makes sense to me. I was thinking in the lines of having an internal only storage that is exposed to the public through an encapsulating type. However, I didn't mention that at all and I am not even sure there is enough benefit to the indirection.

---

I think the case I should have brought up as concrete example is the mutex usage in `insertEvents` & `getStoredQueue`. In this case, we are locking the mutex, reading the disk, releasing the lock, then locking it immediately after to persist objects and then finally releasing it again. This type of relationship may be necessary in some cases when part of the relationship is outside of our control. However, in this case, we are able to control this, so I think it'd be good to lock the mutex once, read the events, persist the objects, and then release the lock. Here is what I am thinking:

```
    private fun getInternalStoredQueue(): ArrayList<Map<String, Any?>?> {
        var storedQueue: ArrayList<Map<String, Any?>?> = ArrayList()
        try {
            val fis = context.applicationContext.openFileInput(STORAGE_KEY)
            val ois = ObjectInputStream(fis)
            @Suppress("UNCHECKED_CAST")
            storedQueue = ois.readObject() as ArrayList<Map<String, Any?>?>
            ois.close()
            fis.close()
        } catch (ex: EOFException) {
            // Nothing to do here.
        } catch (ex: FileNotFoundException) {
            // Nothing to do here. Means there was no saved queue.
        } catch (ex: Exception) {
            ParselyTracker.PLog(
                "Exception thrown during queue deserialization: %s",
                ex.toString()
            )
        }
        return storedQueue
    }

    override suspend fun getStoredQueue(): ArrayList<Map<String, Any?>?> = mutex.withLock {
        return getInternalStoredQueue()
    }

```

To connect the dots, this setup is actually pretty similar to what I mentioned related to "internal only storage" above. Instead of having a second type to separate the internal functions, we are utilizing private functions. Although, If we did have an internal only type that doesn't have a mutex, we'd have been able to use that one in other parts of our internal implementation and avoid the mutex inside another mutex type of situation. It's also generally a design that I like, but this is definitely outside of this PR and I am just sharing my general thoughts on the subject.

Looking forward to hear your thoughts!

Reply:
> I think it'd be good to lock the mutex once, read the events, persist the objects, and then release the lock

True - I see your point. I've updated the code in c32302f to lock the mutex once - it seems even safer this way, as e.g. in `insertEvent`, between `getStoredEvents` and `mutex.withLock` a new event *might* be added to the queue in a different lock, making the final list provided to `persistObject` incorrect. Thanks!

> If we did have an internal only type that doesn't have a mutex, we'd have been able to use that one in other parts of our internal implementation and avoid the mutex inside another mutex type of situation.

`LocalStorageRepository` is internal (in a sense that it's impossible for SDK consumer to trigger any methods on it). Do you think about a different definition of "internal" here? 

Thinking about removing the necessity of mutex nesting, I don't see a clear solution with file-based storage. At a time of querying or saving data, we can't allow for other operations to be performed on the file and there's a constant possibility that e.g. `InMemoryBuffer` will try to save new events.

But I can also be tunnel-vissioned about the proposed approach. I think it'd be more clear for me if we discussed it with some code snippets, maybe in context of #94 ?

Reply:
I think we have made some good improvements to the Mutex usage and I am fine with leaving it as is to keep the more important parts of the project moving.

> Thinking about removing the necessity of mutex nesting, I don't see a clear solution with file-based storage. At a time of querying or saving data, we can't allow for other operations to be performed on the file and there's a constant possibility that e.g. `InMemoryBuffer` will try to save new events.

I think in the ideal scenario, all writes for our storage would go through a single mutex. So, no 2 things could write to it at once. We'd also only let someone read the storage if there is no write access happening at the same time. This would be much closer to how a real DB works. Having said that, it's quite an involved process to get this right and I am not sure if there is enough benefit in trying to optimize it to the fullest.

If the current solution works out fine and doesn't have any major performance concerns, keeping it simpler seems like the way to go for us.